File: python/x402/src/x402_tron/utils/gasfree.py
"""
GasFree utility functions for address calculation, TIP-712 hashing, and HTTP API interaction.
"""

import hmac
import logging
import time
from typing import Any, Dict, Optional

import httpx

from x402_tron.utils.address import tron_address_to_evm

logger = logging.getLogger(__name__)

# GasFree TIP-712 Types
GASFREE_PERMIT_TRANSFER_TYPES = {
    "PermitTransfer": [
        {"name": "token", "type": "address"},
        {"name": "serviceProvider", "type": "address"},
        {"name": "user", "type": "address"},
        {"name": "receiver", "type": "address"},
        {"name": "value", "type": "uint256"},
        {"name": "maxFee", "type": "uint256"},
        {"name": "deadline", "type": "uint256"},
        {"name": "version", "type": "uint256"},
        {"name": "nonce", "type": "uint256"},
    ]
}

GASFREE_API_BASE_URL = "https://open.gasfree.io/tron"


class GasFreeAPIClient:
    """Official GasFree HTTP API client implementation"""

    def __init__(
        self, base_url: str, api_key: Optional[str] = None, api_secret: Optional[str] = None
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.api_secret = api_secret

    def _generate_signature(self, method: str, path: str, timestamp: int) -> str:
        """Generate HMAC-SHA256 signature for API authentication"""
        if not self.api_secret:
            return ""

        # String to sign: {method}{path}{timestamp}
        msg = f"{method.upper()}{path}{timestamp}"
        import base64

        signature_bytes = hmac.new(
            self.api_secret.encode("utf-8"), msg.encode("utf-8"), digestmod="sha256"
        ).digest()
        return base64.b64encode(signature_bytes).decode("utf-8")

    def _get_headers(self, method: str, path: str) -> Dict[str, str]:
        """Generate headers with authentication"""
        headers = {"Content-Type": "application/json"}
        if not self.api_key or not self.api_secret:
            return headers

        timestamp = int(time.time())
        signature = self._generate_signature(method, path, timestamp)

        headers.update(
            {"Timestamp": str(timestamp), "Authorization": f"ApiKey {self.api_key}:{signature}"}
        )
        return headers

    async def get_nonce(self, user: str, token: str, chain_id: int) -> int:
        """Get the current recommended nonce for a user account
        Official endpoint: GET /api/v1/address/{accountAddress}
        """
        data = await self.get_address_info(user)
        return int(data.get("nonce", 0))

    async def get_address_info(self, user: str) -> Dict[str, Any]:
        """Get account info (activation, balance, nonce) for a user
        Official endpoint: GET /api/v1/address/{accountAddress}
        """
        path = f"/api/v1/address/{user}"
        async with httpx.AsyncClient() as client:
            url = f"{self.base_url}{path}"
            try:
                headers = self._get_headers("GET", path)
                response = await client.get(url, headers=headers)
                response.raise_for_status()
                result = response.json()
                if result.get("code") != 200:
                    raise RuntimeError(
                        f"API error: {result.get('message') or result.get('reason')}"
                    )

                return result.get("data", {})
            except Exception as e:
                logger.error(f"Failed to get address info from GasFree API: {e}")
                raise

    async def submit(self, domain: Dict[str, Any], message: Dict[str, Any], signature: str) -> str:
        """Submit a signed GasFree transaction to the official relayer
        Official endpoint: POST /api/v1/gasfree/submit
        """
        path = "/api/v1/gasfree/submit"
        async with httpx.AsyncClient() as client:
            url = f"{self.base_url}{path}"

            # 官方 API 提交格式要求（基于文档 3.3 章节）
            # 注意：sig 参数不带 0x 前缀
            sig = signature[2:] if signature.startswith("0x") else signature
            payload = {
                "token": message["token"],
                "serviceProvider": message["serviceProvider"],
                "user": message["user"],
                "receiver": message["receiver"],
                "value": str(message["value"]),
                "maxFee": str(message["maxFee"]),
                "deadline": int(message["deadline"]),
                "version": int(message["version"]),
                "nonce": int(message["nonce"]),
                "sig": sig,
                "requestId": f"x402-{int(time.time())}-{sig[:8]}",
            }

            try:
                headers = self._get_headers("POST", path)
                response = await client.post(url, json=payload, headers=headers)
                response.raise_for_status()
                result = response.json()
                if result.get("code") != 200:
                    raise RuntimeError(
                        f"API error: {result.get('message') or result.get('reason')}"
                    )

                data = result.get("data", {})
                return data.get("id")  # 返回 traceId
            except Exception as e:
                logger.error(f"Failed to submit GasFree transaction: {e}")
                raise


def get_gasfree_domain(chain_id: int, verifying_contract: str) -> Dict[str, Any]:
    """Get GasFree TIP-712 domain"""
    return {
        "name": "GasFreeController",
        "version": "V1.0.0",
        "chainId": chain_id,
        "verifyingContract": tron_address_to_evm(verifying_contract),
    }

File: typescript/packages/x402/src/utils/gasfree.ts
import { GASFREE_PRIMARY_TYPE, GASFREE_DOMAIN_TYPE } from '../abi.js';

/**
 * GasFree utility functions for API interaction and domain helpers.
 */

export interface GasFreeResponse<T> {
  code: number;
  reason: string | null;
  message: string | null;
  data: T;
}

export interface GasFreeAsset {
  tokenAddress: string;
  tokenSymbol: string;
  activateFee: number;
  transferFee: number;
  decimal: number;
  frozen: number;
  balance?: string | number;
}

export interface GasFreeAddressInfo {
  accountAddress: string;
  gasFreeAddress: string;
  active: boolean;
  nonce: number;
  allowSubmit: boolean;
  assets: GasFreeAsset[];
}

export interface GasFreeSubmitResponseData {
  id: string;
  state: 'WAITING' | 'INPROGRESS' | 'CONFIRMING' | 'SUCCEED' | 'FAILED';
  createdAt: string;
  accountAddress: string;
  gasFreeAddress: string;
  providerAddress: string;
  targetAddress: string;
  nonce: number;
  tokenAddress: string;
  amount: string;
  expiredAt: string;
  reason?: string;
  txnHash?: string;
  txnState?: 'INIT' | 'NOT_ON_CHAIN' | 'ON_CHAIN' | 'SOLIDITY' | 'ON_CHAIN_FAILED';
}

export interface GasFreeProvider {
  address: string;
  name: string;
  icon: string;
  website: string;
  config: {
    maxPendingTransfer: number;
    minDeadlineDuration: number;
    maxDeadlineDuration: number;
    defaultDeadlineDuration: number;
  };
}

// GasFree TIP-712 Types
export const GASFREE_TYPES = {
  EIP712Domain: GASFREE_DOMAIN_TYPE,
  [GASFREE_PRIMARY_TYPE]: [
    { name: 'token', type: 'address' },
    { name: 'serviceProvider', type: 'address' },
    { name: 'user', type: 'address' },
    { name: 'receiver', type: 'address' },
    { name: 'value', type: 'uint256' },
    { name: 'maxFee', type: 'uint256' },
    { name: 'deadline', type: 'uint256' },
    { name: 'version', type: 'uint256' },
    { name: 'nonce', type: 'uint256' },
  ],
} as const;

export class GasFreeAPIClient {
  private baseUrl: string;
  private apiKey?: string;
  private apiSecret?: string;

  constructor(baseUrl: string, apiKey?: string, apiSecret?: string) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
    this.apiSecret = apiSecret;
  }

  /**
   * Generate HMAC signature for authentication
   */
  private async generateSignature(method: string, path: string, timestamp: number): Promise<string> {
    if (!this.apiSecret) return '';

    const message = `${method.toUpperCase()}${path}${timestamp}`;
    
    // Check if we are in Node.js environment
    if (typeof process !== 'undefined' && process.versions && process.versions.node) {
      const crypto = await import('node:crypto');
      return crypto
        .createHmac('sha256', this.apiSecret)
        .update(message)
        .digest('base64');
    } else {
      // Browser implementation using Web Crypto API
      const encoder = new TextEncoder();
      const keyData = encoder.encode(this.apiSecret);
      const msgData = encoder.encode(message);
      
      const cryptoKey = await globalThis.crypto.subtle.importKey(
        'raw',
        keyData,
        { name: 'HMAC', hash: 'SHA-256' },
        false,
        ['sign']
      );
      
      const signature = await globalThis.crypto.subtle.sign('HMAC', cryptoKey, msgData);
      return btoa(String.fromCharCode(...new Uint8Array(signature)));
    }
  }

  /**
   * Get headers with authentication if keys are provided
   */
  private async getHeaders(method: string, path: string): Promise<Record<string, string>> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.apiKey && this.apiSecret) {
      const timestamp = Math.floor(Date.now() / 1000);
      
      // The Provider expects the FULL path including /nile or /tron for signature
      let fullPath = path;
      const urlParts = this.baseUrl.split('/');
      if (urlParts.length > 3) {
        const prefix = '/' + urlParts.slice(3).join('/');
        if (!fullPath.startsWith(prefix)) {
          fullPath = prefix + path;
        }
      }

      const signature = await this.generateSignature(method, fullPath, timestamp);
      headers['Timestamp'] = timestamp.toString();
      headers['Authorization'] = `ApiKey ${this.apiKey}:${signature}`;
    }

    return headers;
  }

  /**
   * Get all supported service providers
   */
  async getProviders(): Promise<GasFreeProvider[]> {
    const path = '/api/v1/config/provider/all';
    const url = `${this.baseUrl}${path}`;
    const headers = await this.getHeaders('GET', path);
    
    const response = await fetch(url, { headers });
    const bodyText = await response.text();

    if (!response.ok) {
      throw new Error(`GasFree config API error: ${response.status}`);
    }
    const result = JSON.parse(bodyText) as GasFreeResponse<{ providers: GasFreeProvider[] }>;
    if (result.code !== 200) {
      throw new Error(`GasFree config API error: ${result.message || result.reason}`);
    }
    return result.data.providers;
  }

  /**
   * Get the current recommended nonce for a user
   */
  async getNonce(user: string): Promise<number> {
    try {
      const data = await this.getAddressInfo(user);
      return data.nonce ?? 0;
    } catch (error) {
      console.warn(`Failed to get nonce from GasFree API: ${error}. Defaulting to 0.`);
      return 0;
    }
  }

  /**
   * Get full account info (activation, balance, nonce) for a user
   */
  async getAddressInfo(user: string): Promise<GasFreeAddressInfo> {
    const path = `/api/v1/address/${user}`;
    const url = `${this.baseUrl}${path}`;
    const headers = await this.getHeaders('GET', path);

    const response = await fetch(url, { headers });
    const bodyText = await response.text();

    if (!response.ok) {
      console.error(`GasFree API HTTP error ${response.status}: ${bodyText}`);
      throw new Error(`GasFree API HTTP error: ${response.status} - Body: ${bodyText}`);
    }
    const result = JSON.parse(bodyText) as GasFreeResponse<GasFreeAddressInfo>;
    if (result.code !== 200) {
      throw new Error(`GasFree API error: ${result.message || result.reason} - Body: ${bodyText}`);
    }
    return result.data;
  }

  /**
   * Get status of a submitted GasFree transaction
   */
  async getStatus(traceId: string): Promise<GasFreeSubmitResponseData> {
    const path = `/api/v1/gasfree/${traceId}`;
    const url = `${this.baseUrl}${path}`;
    const headers = await this.getHeaders('GET', path);
    const response = await fetch(url, { headers });
    const bodyText = await response.text();

    if (!response.ok) {
      throw new Error(`GasFree status API error: ${response.status}`);
    }
    const result = JSON.parse(bodyText) as GasFreeResponse<GasFreeSubmitResponseData>;
    if (result.code !== 200) {
      throw new Error(`GasFree status API error: ${result.message || result.reason}`);
    }
    return result.data;
  }

  /**
   * Wait for a GasFree transaction to reach a terminal state or acceptable timeout state
   */
  async waitForSuccess(
    traceId: string,
    timeout: number = 180000,
    pollInterval: number = 5000
  ): Promise<GasFreeSubmitResponseData> {
    const startTime = Date.now();
    let statusData: GasFreeSubmitResponseData | undefined;

    while (Date.now() - startTime < timeout) {
      statusData = await this.getStatus(traceId);
      const state = statusData.state.toUpperCase();
      const txnState = (statusData.txnState || '').toUpperCase();

      // 1. Immediate return for final states
      if (state === 'SUCCEED') {
        return statusData;
      }
      if (state === 'FAILED' || txnState === 'ON_CHAIN_FAILED') {
        throw new Error(`GasFree transaction failed. Reason: ${statusData.reason || 'Unknown'}`);
      }

      console.debug(`GasFree transaction ${traceId} is ${state} (${txnState}), waiting...`);
      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    // 2. Timeout reached
    if (statusData) {
      const finalState = statusData.state.toUpperCase();
      const finalTxnState = (statusData.txnState || '').toUpperCase();

      if (finalState === 'CONFIRMING' && finalTxnState === 'ON_CHAIN') {
        console.info(`GasFree transaction ${traceId} reached CONFIRMING/ON_CHAIN at timeout. Success.`);
        return statusData;
      }
    }

    throw new Error(`GasFree transaction ${traceId} timed out after ${timeout / 1000}s`);
  }

  /**
   * Submit a signed GasFree transaction to the official relayer
   */
  async submit(domain: any, message: any, signature: string): Promise<string> {
    const path = '/api/v1/gasfree/submit';
    const url = `${this.baseUrl}${path}`;
    
    const payload = {
      token: message.token,
      serviceProvider: message.serviceProvider,
      user: message.user,
      receiver: message.receiver,
      value: message.value.toString(),
      maxFee: message.maxFee.toString(),
      deadline: Number(message.deadline),
      version: 1,
      nonce: Number(message.nonce),
      sig: signature.startsWith('0x') ? signature.slice(2) : signature,
      requestId: `x402-${Date.now()}-${signature.slice(2, 10)}`,
    };

    try {
      const headers = await this.getHeaders('POST', path);
      const response = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(payload),
      });

      const bodyText = await response.text();

      if (!response.ok) {
        console.error(`GasFree submit HTTP error ${response.status}: ${bodyText}`);
        throw new Error(`GasFree submit HTTP error: ${response.status} - Body: ${bodyText}`);
      }

      const result = JSON.parse(bodyText) as GasFreeResponse<GasFreeSubmitResponseData>;
      if (result.code !== 200) {
        throw new Error(`GasFree submit API error: ${result.message || result.reason} - Body: ${bodyText}`);
      }
      return result.data.id;
    } catch (error) {
      console.error(`Failed to submit GasFree transaction: ${error}`);
      throw error;
    }
  }
}
